com.unity.shadergraph/Editor/Drawing/MaterialEditor/ShaderGraphPropertyDrawers.cs

```diff
@@ -0,0 +1,256 @@
+using System;
+using System.Collections.Generic;
+using UnityEditor;
+using UnityEngine;
+using UnityEngine.Rendering;
+using Object = UnityEngine.Object;
+
+namespace UnityEditor.ShaderGraph.Drawing
+{
+    internal static class ShaderGraphPropertyDrawers
+    {
+
+        public static void DrawShaderGraphGUI(MaterialEditor materialEditor, MaterialProperty[] properties)
+        {
+            Material m = materialEditor.target as Material;
+            Shader s = m.shader;
+            string path = AssetDatabase.GetAssetPath(s);
+            ShaderGraphMetadata metadata = null;
+            foreach (var obj in AssetDatabase.LoadAllAssetsAtPath(path))
+            {
+                if(obj is ShaderGraphMetadata meta)
+                {
+                    metadata = meta;
+                    break;
+                }
+            }
+            Debug.Assert(metadata != null, "Cannot draw ShaderGraph GUI on a non-ShaderGraph material", materialEditor.target);
+            DrawShaderGraphGUI(materialEditor, properties, metadata.categoryDatas);
+        }
+
+        private static Rect GetRect(MaterialProperty prop)
+        {
+            return EditorGUILayout.GetControlRect(true, MaterialEditor.GetDefaultPropertyHeight(prop));
+        }
+        private static Rect GetRect()
+        {
+            return EditorGUILayout.GetControlRect();
+        }
+
+        private static MaterialProperty FindProperty(string propertyName, MaterialProperty[] properties)
+        {
+            foreach(var prop in properties)
+            {
+                if(prop.name == propertyName)
+                {
+                    return prop;
+                }
+            }
+            throw new ArgumentException("no property was found with the name " + propertyName);
+        }
+
+        public static void DrawShaderGraphGUI(MaterialEditor materialEditor, MaterialProperty[] properties, IEnumerable<MinimalCategoryData> categoryDatas)
+        {
+            foreach(MinimalCategoryData mcd in categoryDatas)
+            {
+                DrawCategory(materialEditor, properties, mcd);
+            }
+
+            EditorGUILayout.Space();
+            EditorGUILayout.Space();
+            if (SupportedRenderingFeatures.active.editableMaterialRenderQueue)
```

Comment:
Always use curlybois please, saves bugs later when someone doesn't notice they're not there and adds a line here.